services/director-v2/src/simcore_service_director_v2/utils/distributed_identifier.py

```diff
@@ -0,0 +1,281 @@
+        _logger.info("Will remove unused  %s", list(tracked_data.keys()))
+
+        for identifier, cleanup_context in tracked_data.items():
+            if await self.is_used(identifier, cleanup_context):
+                continue
+
+            await self.remove(identifier)
+
+    async def create(
+        self, *, cleanup_context: CleanupContext, **extra_kwargs
+    ) -> tuple[Identifier, ResourceObject]:
+        """Used for creating the resources
+
+        Arguments:
+            cleanup_context -- user defined CleanupContext object
+            **extra_kwargs -- can be overloaded by the user
+
+        Returns:
+            tuple[identifier for the resource, resource object]
+        """
+        identifier, result = await self._create(**extra_kwargs)
+        await self.redis_client_sdk.redis.set(
+            self._to_redis_key(identifier),
+            self._serialize_cleanup_context(cleanup_context),
+        )
+        return identifier, result
+
+    async def remove(self, identifier: Identifier, *, reraise: bool = False) -> None:
+        """Attempts to remove the resource, if an error occurs it is logged.
+
+        Arguments:
+            identifier -- user chosen identifier for the resource
+            reraise -- when True raises any exception raised by ``destroy`` (default: {False})
+        """
+
+        cleanup_context = await self._get_identifier_context(identifier)
+        if cleanup_context is None:
+            _logger.warning(
+                "Something went wrong, did not find any context for %s", identifier
+            )
+            return
+
+        with log_context(
+            _logger, logging.DEBUG, f"{self.__class__}: removing {identifier}"
+        ), log_catch(_logger, reraise=reraise):
+            await self._destroy(identifier, cleanup_context)
+
+        await self.redis_client_sdk.redis.delete(self._to_redis_key(identifier))
+
+    @classmethod
+    @abstractmethod
+    def _deserialize_identifier(cls, raw: str) -> Identifier:
+        """User provided deserialization for the identifier
+
+        Arguments:
+            raw -- stream to be deserialized
+
+        Returns:
+            an identifier object
+        """
+
+    @classmethod
+    @abstractmethod
+    def _serialize_identifier(cls, identifier: Identifier) -> str:
+        """User provided serialization for the identifier
+
+        Arguments:
+            cleanup_context -- user defined identifier object
+
+        Returns:
+            object encoded as string
+        """
+
+    @classmethod
+    @abstractmethod
+    def _deserialize_cleanup_context(cls, raw: str) -> CleanupContext:
+        """User provided deserialization for the context
+
+        Arguments:
+            raw -- stream to be deserialized
+
+        Returns:
+            an object of the type chosen by the user
+        """
+
+    @classmethod
+    @abstractmethod
+    def _serialize_cleanup_context(cls, cleanup_context: CleanupContext) -> str:
```

Comment:
So here there are three entities: an identifier, a cleanup-context and the resource object. 
You documented that the first and last are serialised in redis. here I see only serialisation/deserialization for the first two only. 

Here I see serialization/deserialization abstract methods for two of them but not for the resource object?

Reply:
My bad. I've fixed the comment. The code is correct. You only store in the DB the identifier and the cleanup-context. Only these are required for object retrieval and removal.
The `ResourceObject` can always be retrieved via its identifier stored in Redis.